src/V3Fork.cpp

```diff
@@ -34,12 +46,351 @@
+        // TODO: Ensure no collisions occur
+        string n = "__VDynScope__" + (!fromp->name().empty() ? (fromp->name() + "__") : "ANON__")
+                   + cvtToHex(fromp);
+        return n;
+    }
+
+    static string generateDynScopeHandleName(const AstNode* fromp) {
+        return "__VDynScope_" + (fromp->name().empty() ? cvtToHex(fromp) : fromp->name());
+    }
+
+    AstNode* getProcStmts() {
+        AstNode* stmtsp = nullptr;
+        if (!m_procp) return nullptr;
+        if (AstBegin* beginp = VN_CAST(m_procp, Begin))
+            stmtsp = beginp->stmtsp();
+        else if (AstNodeFTask* taskp = VN_CAST(m_procp, NodeFTask))
+            stmtsp = taskp->stmtsp();
+        else
+            v3fatal("m_procp is not a begin block or a procedure");
+        return stmtsp;
+    }
+
+    AstNodeModule* m_modp;
+    AstNode* m_procp;
+    std::set<AstVar*> m_captures;
+    DynScopeInstance m_instance;
+};
+
+//######################################################################
+// Dynamic scope visitor, creates classes and objects for dynamic scoping of variables and
+// replaces references to varibles that need a dynamic scope with references to object's
+// members
+
+class DynScopeVisitor final : public VNVisitor {
+private:
+    // NODE STATE
+    // AstVar::user1()          -> int, timing-control fork nesting level of that variable
+    // AstVarRef::user2()       -> bool, 1 = Node is a class handle reference. The handle gets
+    //                                       modified in the context of this reference.
+    const VNUser1InUse m_inuser1;
+    const VNUser2InUse m_inuser2;
+
+    // STATE
+    AstNodeModule* m_modp = nullptr;  // Module we are currently under
+    AstNode* m_procp = nullptr;  // Function/task/block we are currently under
+    std::map<AstNode*, DynScopeFrame*> m_frames;  // Mapping from nodes to related DynScopeFrames
+    VMemberMap m_memberMap;  // Class member look-up
+    int m_forkDepth = 0;  // Number of asynchronous forks we are currently under
+    bool m_afterTimingControl = false;  // A timing control might've be executed in the current
+                                        // process
+
+    // METHODS
+
+    DynScopeFrame* frameOf(AstNode* nodep) {
+        auto frameIt = m_frames.find(nodep);
+        if (frameIt == m_frames.end()) return nullptr;
+        return frameIt->second;
+    }
+
+    const DynScopeFrame* frameOf(AstNode* nodep) const {
+        auto frameIt = m_frames.find(nodep);
+        if (frameIt == m_frames.end()) return nullptr;
+        return frameIt->second;
+    }
+
+    DynScopeFrame* pushDynScopeFrame() {
+        DynScopeFrame* const frame = new DynScopeFrame{m_modp, m_procp};
+        auto r = m_frames.emplace(std::make_pair(m_procp, frame));
+        UASSERT_OBJ(r.second, m_modp, "Procedure already contains a frame");
+        return frame;
+    }
+
+    static string generateDynScopeClassName(AstNode* fromp) {
+        // TODO: Ensure no collisions occur
```

Comment:
```suggestion
        // FIXME: Ensure no collisions occur
```

I think we shouldn't merge until that's done - FIXME will ensure that (via test failure).

Reply:
Actually, I think it's basically impossible to cause a collision here. The user would have to somehow predict the name of the generated name, not sure if they could also get around the `__V` prefix. The generated names can't collide with each other.

Reply:
Thinking about it , I don't recall verilator doing the same kind of checks when generating variables such as `__Vincrement#`, so I think it's fair to assume no collision here,